x-pack/plugins/security_solution/server/lib/siem_migrations/rules/task/agent/nodes/match_prebuilt_rule/match_prebuilt_rule.ts

```diff
@@ -7,55 +7,56 @@
 
 import { JsonOutputParser } from '@langchain/core/output_parsers';
 import { SiemMigrationRuleTranslationResult } from '../../../../../../../../common/siem_migrations/constants';
+import type { RuleMigrationsRetriever } from '../../../retrievers';
 import type { ChatModel } from '../../../util/actions_client_chat';
-import { filterPrebuiltRules, type PrebuiltRulesMapByName } from '../../../util/prebuilt_rules';
 import type { GraphNode } from '../../types';
 import { MATCH_PREBUILT_RULE_PROMPT } from './prompts';
 
 interface GetMatchPrebuiltRuleNodeParams {
   model: ChatModel;
-  prebuiltRulesMap: PrebuiltRulesMapByName;
+  ruleMigrationsRetriever: RuleMigrationsRetriever;
 }
 
 interface GetMatchedRuleResponse {
   match: string;
 }
 
 export const getMatchPrebuiltRuleNode =
-  ({ model, prebuiltRulesMap }: GetMatchPrebuiltRuleNodeParams): GraphNode =>
+  ({ model, ruleMigrationsRetriever }: GetMatchPrebuiltRuleNodeParams): GraphNode =>
   async (state) => {
-    const mitreAttackIds = state.original_rule.mitre_attack_ids;
-    if (!mitreAttackIds?.length) {
-      return {};
-    }
-
-    const filteredPrebuiltRulesMap = filterPrebuiltRules(prebuiltRulesMap, mitreAttackIds);
-    if (filteredPrebuiltRulesMap.size === 0) {
-      return {};
-    }
+    const query = state.semantic_query;
+    const techniqueIds = state.original_rule.mitre_attack_ids || [];
+    const prebuiltRules = await ruleMigrationsRetriever.prebuiltRules.getRules(
+      query,
+      techniqueIds.join(',')
+    );
 
     const outputParser = new JsonOutputParser();
     const matchPrebuiltRule = MATCH_PREBUILT_RULE_PROMPT.pipe(model).pipe(outputParser);
 
-    const elasticSecurityRules = [...filteredPrebuiltRulesMap.keys()].join('\n');
+    const elasticSecurityRules = prebuiltRules.map((rule) => {
+      return {
+        name: rule.name,
+        description: rule.description,
+      };
+    });
+
     const response = (await matchPrebuiltRule.invoke({
-      elasticSecurityRules,
+      rules: JSON.stringify(elasticSecurityRules, null, 2),
       ruleTitle: state.original_rule.title,
     })) as GetMatchedRuleResponse;
     if (response.match) {
-      const result = filteredPrebuiltRulesMap.get(response.match);
-      if (result != null) {
+      const matchedRule = prebuiltRules.find((r) => r.name === response.match);
+      if (matchedRule) {
         return {
           elastic_rule: {
-            title: result.rule.name,
-            description: result.rule.description,
-            prebuilt_rule_id: result.rule.rule_id,
-            id: result.installedRuleId,
```

Comment:
are we planning to set the `id` for those rules that matched prebuilt rule that has been already installed?

Reply:
I thought an earlier meeting we mentioned we did not need that field? The return is only the matching prebuilt rule, should it matter if its already installed? I thought we wanted to check that in the UI?

Reply:
Yes, we should set the id if it is already installed, we'll try to keep that consistent from the UI, but better if we already define it here as well.

Reply:
I see we lost that information with the introduction of RAG. But, since we index the documents on every run, would it make sense to store the `installedRuleId` in the index as well?

Reply:
Adding it is no problem, I removed it only because I thought we did not want it, adding it in is straightforward

Reply:
Will commit the change soon, tested by installing all prebuilt rules and running again. Langsmith now returns:

```
{
  "elastic_rule": {
    "title": "Linux User Account Creation",
    "description": "Identifies attempts to create new users. Attackers may add new users to establish persistence on a system.",
    "id": "d54a0a03-9464-499f-95a6-a28dbe6f0f22",
    "prebuilt_rule_id": "edfd5ca9-9d6c-44d9-b615-1e56b920219c"
  },
  "translation_result": "full"
}
```

Reply:
Storage in RAG index:
```
{
"_index": ".kibana-siem-rule-migrations-prebuiltrules-default",
        "_id": "015cca13-8832-49ac-a01b-a396114809f6",
        "_score": 0,
        "_source": {
          "rule_id": "015cca13-8832-49ac-a01b-a396114809f6",
          "@timestamp": "2024-12-05T21:42:24.284Z",
          "name": "AWS Redshift Cluster Creation",
          "description": "Identifies the creation of an Amazon Redshift cluster. Unexpected creation of this cluster by a non-administrative user may indicate a permission or role issue with current users. If unexpected, the resource may not properly be configured and could introduce security vulnerabilities.",
          "installedRuleId": "96b921c0-ff31-4bfc-8089-a49516156605"
        }
      }
```